include/graphblas/algorithms/amg/multigrid_v_cycle.hpp

```diff
@@ -0,0 +1,327 @@
+				RC ret = SUCCESS;
+				ret = ret ? ret : set( coarsening_data.Ax_finer, 0 );
+
+				ret = ret ? ret : grb::mxv< grb::descriptors::transpose_matrix >(
+					coarsening_data.Ax_finer,
+					coarsening_data.coarsening_matrix,
+					coarsening_data.z,
+					ring
+				);
+				assert( ret == SUCCESS );
+				// x_fine += Ax_finer;
+				ret = ret ? ret : grb::foldl( x_fine, coarsening_data.Ax_finer, ring.getAdditiveMonoid() );
+				assert( ret == SUCCESS );
+				return ret;
+			}
+
+			template< typename IOType, typename NonzeroType, class Ring >
+			grb::RC spai0_smoother( system_data< IOType, NonzeroType > &data, const Ring &ring ) {
+				RC ret = SUCCESS;
+
+				NonzeroType alpha = 1.;
+				ret = ret ? ret : grb::eWiseMulAdd( data.z, alpha, data.A_diagonal, data.r, ring );
+				assert( ret == SUCCESS );
+				return ret;
+			}
+
+			/**
+			 * Runs \p smoother_steps iteration of the SPAI0 smoother,
+			 *        with inputs and outputs stored inside \p data.
+			 *
+			 * @tparam IOType type of result and intermediate vectors used during computation
+			 * @tparam NonzeroType type of matrix values
+			 * @tparam Ring the ring of algebraic operators zero-values
+			 *
+			 * @param[in,out] data \ref system_data data structure with relevant inpus
+			 *                 and outputs: system matrix, initial solution, residual,
+			 *                 system matrix colors, temporary vectors
+			 * @param[in] smoother_steps how many smoothing steps to run
+			 * @param[in] ring the ring to perform the operations on
+			 * @return grb::RC::SUCCESS if the algorithm could correctly terminate,
+			 *         the error code of the first unsuccessful operation otherwise
+			 */
+			template< typename IOType, typename NonzeroType, class Ring, class Minus >
+			grb::RC run_spai0_smoother( system_data< IOType, NonzeroType > &data,
+								  const size_t smoother_steps, const Ring &ring, const Minus &minus ) {
+				RC ret = SUCCESS;
+
+				for( size_t i = 0; i < smoother_steps && ret == SUCCESS; i++ ) {
+					ret = ret ? ret : grb::set( data.smoother_temp, 0 );
+					ret = ret ? ret : grb::mxv( data.smoother_temp, data.A, data.z, ring );
+					ret = ret ? ret : grb::eWiseApply( data.smoother_temp, data.r, data.smoother_temp,
+													   minus );
+
+#ifdef HPCG_PRINT_STEPS
+					std::cout << " data.A(spai0): " << nrows(data.A) << " x " << ncols(data.A) << " \n";
+					print_norm( data.A_diagonal, " data.A_diagonal" );
+					print_norm( data.smoother_temp, " data.smoother_temp" );
+					print_norm( data.z, " data.z" );
+#endif
+
+					ret = ret ? ret :
+						grb::eWiseLambda(
+							[ &data ]( const size_t i ) {
+#ifdef HPCG_PRINT_STEPS
+								if( i < 10 || i + 10 > size( data.A_diagonal ) ) {
+									std::cout << " i= " << i
+									          << " data.z[i]= " << data.z[ i ]
+									          << " data.A_diagonal[i]= " << data.A_diagonal[ i ]
+									          << " data.smoother_temp[i]= " << data.smoother_temp[ i ]
+									          << "\n";
```

Comment:
while I don't mind this too much, elsewhere the indentation of `<<` happens through tabs only, just as a note